kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```diff
@@ -0,0 +1,325 @@
+  private val secretName = s"driver-launcher-service-secret-$kubernetesAppId"
+  private val driverLauncherSelectorValue = s"driver-launcher-$launchTime"
+
+  private val secretBytes = new Array[Byte](128)
+  SECURE_RANDOM.nextBytes(secretBytes)
+  private val secretBase64String = Base64.encodeBase64String(secretBytes)
+  private implicit val retryableExecutionContext = ExecutionContext
+    .fromExecutorService(
+      Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+        .setNameFormat("kubernetes-client-retryable-futures-%d")
+        .setDaemon(true)
+        .build()))
+
+  def run(): Unit = {
+    var k8ConfBuilder = new ConfigBuilder()
+      .withApiVersion("v1")
+      .withMasterUrl(clientArgs.kubernetesMaster)
+      .withNamespace(namespace)
+    clientArgs.kubernetesCaCertFile.foreach { f => k8ConfBuilder = k8ConfBuilder.withCaCertFile(f)}
+    clientArgs.kubernetesClientKeyFile.foreach(file =>
+      k8ConfBuilder = k8ConfBuilder.withClientKeyFile(file))
+    clientArgs.kubernetesClientCertFile.foreach(file =>
+        k8ConfBuilder = k8ConfBuilder.withClientCertFile(file))
+
+    val k8ClientConfig = k8ConfBuilder.build
+    Utils.tryWithResource(new DefaultKubernetesClient(k8ClientConfig))(kubernetesClient => {
+      val secret = kubernetesClient.secrets().createNew()
+        .withNewMetadata()
+          .withName(secretName)
+          .endMetadata()
+        .withData(Map((DRIVER_LAUNCHER_SECRET_NAME, secretBase64String)).asJava)
+        .withType("Opaque")
+        .done()
+      try {
+        val driverConfiguration = buildSparkDriverConfiguration()
+        val selectors = Map(DRIVER_LAUNCHER_SELECTOR_LABEL -> driverLauncherSelectorValue).asJava
+        val uiPort = driverConfiguration
+          .sparkConf
+          .get("spark.ui.port")
+          .map(_.toInt)
+          .getOrElse(DEFAULT_UI_PORT)
+        val (servicePorts, containerPorts) = configurePorts(uiPort)
+
+        val service = kubernetesClient.services().createNew()
+          .withNewMetadata()
+            .withName(kubernetesAppId)
+            .endMetadata()
+          .withNewSpec()
+            .withSelector(selectors)
+            .withPorts(servicePorts.asJava)
+            .endSpec()
+          .done()
+
+        val submitCompletedFuture = SettableFuture.create[Boolean]
+
+        val podWatcher = new Watcher[Pod] {
+          override def eventReceived(action: Action, t: Pod): Unit = {
+            if ((action == Action.ADDED || action == Action.MODIFIED)
+                && t.getStatus.getPhase == "Running"
+                && !submitCompletedFuture.isDone) {
+              t.getStatus
+                .getContainerStatuses
+                .asScala
+                .find(status =>
+                  status.getName == DRIVER_LAUNCHER_CONTAINER_NAME && status.getReady) match {
+                case Some(status) =>
+                  try {
+                    val driverLauncher = getDriverLauncherService(k8ClientConfig)
```

Comment:
Yes we need to basically translate the user dependencies and configuration into a downstream spark-submit call, but this allows a client mode implementation that doesn't need to consider how applications are managed/submitted, and drastically simplifies the cluster mode too.
This is exactly what Mesos cluster scheduler does, and what Standalone cluster mode does this too.